prov/sockets/src/sock_av.c

```diff
@@ -524,9 +524,18 @@ int sock_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
 			ret = -FI_EINVAL;
 			goto err2;
 		}
+		if (!_av->attr.map_addr) {
+			_av->table_hdr = mmap(NULL, table_sz, PROT_READ | PROT_WRITE,
+					MAP_SHARED, _av->shared_fd, 0);
+			_av->attr.map_addr = _av->table_hdr + sizeof(_av->table_hdr);
+			attr->map_addr = _av->attr.map_addr;
+			SOCK_LOG_DBG("Updating map_addr: %p\n");
+		} else {
+			SOCK_LOG_DBG("Using map_addr: %p in mmap\n", _av->attr.map_addr);
+			_av->table_hdr = mmap(_av->attr.map_addr, table_sz, PROT_READ | PROT_WRITE,
+					MAP_SHARED, _av->shared_fd, 0);
```

Comment:
The else part (user passes in map_addr) seems to be incorrect. In this case, the map_addr will point to the table_hdr, not the actual address. We need to offset this like you did for the map_addr=NULL case. Also, what happens if the mmap() fails when it tries to use map_addr?

Instead of attempting to mmap with the user provided map_addr, we could just do mmap() with NULL, and return the table address via map_addr.

Man page: The map_addr may be used by the provider to mmap memory allocated for a shared AV between processes; however, the provider is not required to use the map_addr in this fashion. **The only requirement is that an fi_addr_t returned as part of an fi_av_insert call on one process is usable on another process** which opens an AV of the same name at the same map_addr value.